zerver/lib/narrow.py

```diff
@@ -1259,11 +1304,12 @@ def fetch_messages(
         need_message = True
         need_user_message = True
 
-    query: SelectBase
+    query: Union[SelectBase, Tuple[SelectBase, SelectBase]]
     query, inner_msg_id_col = get_base_query_for_search(
         user_profile=user_profile,
         need_message=need_message,
         need_user_message=need_user_message,
+        anchor_date=anchor_date,
```

Comment:
Why do you need to change the base query? Just add another mapping for operator and it's method in `by_method_map` (check `NarrowBuilder`). And then you can send a query with a new operator type: `near-date`.